### PR TITLE
test(python): add pure-python tests and CI for bytecodeparser

### PR DESCRIPTION
**Pull request by MarcoGorelli**
The bytecodeparser is becoming more complex, and going deep into scary Python internals, and is garnering a bit of attention

I'm a bit worried about subtle differences between Python versions resulting in it no longer working, so I've added a pure-Python test and workflow to check the resulting expressions across all supported Python versions

There's no need to install polars (or anything really, other than pytest and numpy) so these workflows only take 16-19 seconds

### CODE REVIEW

File: py-polars/polars/utils/_construction.py [hunk 1]
```diff
@@ -1410,7 +1410,7 @@ def series_to_pydf(
         schema or series_name, schema_overrides=schema_overrides, n_expected=1
     )
     if schema_overrides:
-        new_dtype = list(schema_overrides.values())[0]
+        new_dtype = next(iter(schema_overrides.values()))
```
Comment (MarcoGorelli) [inline py-polars/polars/utils/_construction.py:1413]: ruff did this

Comment (ritchie46) [reply py-polars/polars/utils/_construction.py:1413]: Smart ruff!

**Comment by alexander-beedie** [comment]
> I've added a pure-Python test and workflow to check the resulting expressions across all supported Python versions

Yup, this is an excellent idea... :)

File: .github/workflows/test-bytecode-parser.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  run:
+    working-directory: py-polars
+
+jobs:
+  ubuntu:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ['3.8', '3.9', '3.10', '3.11']
```
Comment (alexander-beedie) [inline .github/workflows/test-bytecode-parser.yml:29]: What do you think about getting ahead of the game and adding "3.12-dev" too? :)

Comment (MarcoGorelli) [reply .github/workflows/test-bytecode-parser.yml]: great idea, but looks like numpy fails to install there

**Comment by MarcoGorelli** [comment]
allright, finally got it sorted out so that both tests (polars-installed and polars-uninstalled) pass 🥳

**Comment by alexander-beedie** [comment]
> allright, finally got it sorted out so that both tests (polars-installed and polars-uninstalled) pass 🥳

I can see why it's where it is, but is it not possible (at all?) to put `test_cases` in the actual tests directory? Seems _really_ peculiar to have them in the library. Must be a reasonable way to tweak this, hmm...🤔

**Comment by MarcoGorelli** [comment]
you're right, it is in fact possible - thanks for calling me out on this! I've put the test cases in `test_udfs.py`, and then have had to add `test/__init__.py` so that `TEST_CASES` can be imported from `test.test_udfs`

**Comment by alexander-beedie** [comment]
> you're right, it is in fact possible - thanks for calling me out on this! I've put the test cases in `test_udfs.py`, and then have had to add `test/__init__.py` so that `TEST_CASES` can be imported from `test.test_udfs`

Awesome! Looks much better now, and the extra peace of mind this will give us for these low-level shenanigans is great :))

**Review by alexander-beedie — APPROVED** [review]

File: .github/workflows/test-bytecode-parser.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        run: |
+          python -m venv .venv
+          echo "$GITHUB_WORKSPACE/py-polars/.venv/bin" >> $GITHUB_PATH
+
+      - name: Install dependencies
+        run: pip install numpy pytest
+
+      - name: Run tests
+        if: github.ref_name != 'main'
+        run: PYTHONPATH=polars/utils pytest tests/test_udfs.py
```
Comment (MarcoGorelli) [inline .github/workflows/test-bytecode-parser.yml:49]: I thought of reporting coverage here, although these tests don't (and can't, as polars isn't installed) hit the entire file, so let's leave that to the main CI job

Comment (alexander-beedie) [reply .github/workflows/test-bytecode-parser.yml:49]: Yup; and they won't actually increase the coverage anyway - let's keep these ones lean/focused :)

**Pull request merged**